### ISSUE TITLE
Double sums are not created or evaluated in the correct order in evalf and lambdify

### ISSUE THREAD

**Issue by leonhardpetzel** [sympy/sympy#27112]
To demonstrate, I create a simple double sum:

```
from sympy import lambdify, abc, Sum

expr = Sum(Sum(1, (abc.j, 0, abc.i)), (abc.i, 0, 10))
fun = lambdify((), expr)
print(fun.__doc__)
fun()
```

Output is :
```
...
Expression:
Sum(1, (j, 0, i), (i, 0, 10))
Source code:
def _lambdifygenerated():
    return (builtins.sum(1 for j in range(0, i+1) for i in range(0, 10+1)))
...
```

And the code throws an error, because the python code should be:

```
def _lambdifygenerated():
    return (builtins.sum(1 for i in range(0, 10+1) for j in range(0, i+1)))
```

So, either the translation of two individual Sums into one Sum produces the wrong order. Or the lamdification order needs to be swapped.

**Comment by bjodah** [comment]
Have you tried applying `.doit()` on your expression?
```python
>>> sympy.Sum(1, (abc.i, 0, 10), (abc.j, 0, abc.i)).doit()
11*i + 11

>>> sympy.Sum(1, (abc.j, 0, abc.i), (abc.i, 0, 10)).doit()
66
```
Looks like `sympy.Sum` expects summations to be specified "from the inside out"? 

`lambdify` still does not like this expression though.

**Comment by leonhardpetzel** [comment]
Yes, it seems like .doit() expects "from the inside out", while lambdify expects "from the outside in".

Further, the implementation for .evalf() (which calls evalf_sum) does not handle nested sums.

**Comment by bjodah** [comment]
It's not only `.doit()` that expects this order, the pretty printing also makes a case for it:
```python
>>> Sum(x, (i, 1, j), (j, 1, 5))
  5     j    
 ___   ___   
 ╲     ╲     
  ╲     ╲    
  ╱     ╱   x
 ╱     ╱     
 ‾‾‾   ‾‾‾   
j = 1 i = 1 
```
As for `.evalf()`, is it reasonable to expect the user to call `.doit()` before `.evalf()`?

So I guess the documentation of `Sum` should be updated to clarify the expectation on this order, and the code-printer class should be fixed (so `lambdify` starts working).

**Comment by leonhardpetzel** [comment]
> As for `.evalf()`, is it reasonable to expect the user to call `.doit()` before `.evalf()`?

I cannot judge this. I expected .evalf() to handle finite sums.
To me it is not obvious, when to use `.doit(), .simplify() or .evalf()` in general. I always considered `.doit()` like a weaker version of simplify. But in this example, `.doit()` works and `.simplify()` "fails".
`sympy.Sum(1, (abc.j, 0, abc.i), (abc.i, 0, 10)).simplify()` returns the expression as is. That is because for Sum, there is a specialization (`_eval_simplify()`), which prevents other evaluation strategies.
Also, `sympy.Product(2, (abc.j, 0, abc.i), (abc.i, 0, 3))` requires a `.doit()` before `.evalf()`. But here, a `.simplify()` works.

> So I guess the documentation of `Sum` should be updated to clarify the expectation on this order, and the code-printer class should be fixed (so `lambdify` starts working).

Agree. I want to mention, that there are multiple occations of limits like Integrals or Products with ExprWithLimits as the common super class.

**Comment by oscarbenjamin** [comment]
> It's not only `.doit()` that expects this order, the pretty printing also makes a case for it:

The order is definitely inside-out. The same is also for integrals:
```python
In [2]: Integral(1, (x, 0, 1), (y, 0, 1))
Out[2]: 
1 1        
⌠ ⌠        
⎮ ⎮ 1 dx dy
⌡ ⌡        
0 0
```

> As for `.evalf()`, is it reasonable to expect the user to call `.doit()` before `.evalf()`?

No, because `.doit` is for symbolic evaluation whereas `.evalf` is for numerical evaluation. Many integrals/summations can be evaluated numerically but not symbolically:
```python
In [4]: s = Sum(exp(-exp(x)), (x, 0, oo))

In [5]: s
Out[5]: 
 ∞        
____      
╲         
 ╲        
  ╲      x
  ╱    -ℯ 
 ╱    ℯ   
╱         
‾‾‾‾      
x = 0     

In [6]: s.doit()
Out[6]: 
 ∞        
____      
╲         
 ╲        
  ╲      x
  ╱    -ℯ 
 ╱    ℯ   
╱         
‾‾‾‾      
x = 0     

In [7]: s.evalf()
Out[7]: 0.434485457898265
```
I'm not sure that evalf has any support for double summations though. See gh-25621 for double and triple integrals.

> So I guess the documentation of `Sum` should be updated to clarify the expectation on this order

The docstring for Sum could use a lot of improvement I think:
https://docs.sympy.org/latest/modules/concrete.html#sympy.concrete.summations.Sum
It uses confusing notation and talks about Karr's definition of a sum. The Karr definition defines sums as being based on a half-open range e.g.:
```
sum_karr(f(x), x, a, b) = f(a) + f(a+1) + ... + f(b-1)
```
The SymPy Sum is
```
sum_sympy(f(x), x, a, b) = f(a) + f(a+1) + ... + f(b-1) + f(b)
```
Essentially then
```
sum_sympy(f(x), x, a, b) = sum_karr(f(x), x, a, b+1)
```
This leads to a confusion around backwards summation:
```python
In [15]: summation(f(x), (x, 0, 2))
Out[15]: f(0) + f(1) + f(2)

In [19]: summation(f(x), (x, 0, -2))
Out[19]: -f(-1)
```
In the Karr notation this makes more sense because:
```python
In [25]: sum_karr = lambda f, x, a, b: summation(f, (x, a, b-1))

In [26]: sum_karr(f(x), x, 0, 2)
Out[26]: f(0) + f(1)

In [27]: sum_karr(f(x), x, 0, -2)
Out[27]: -f(-2) - f(-1)
```
The identities that Karr wants are like:
```python
In [28]: sum_karr(f(x), x, 0, 2)
Out[28]: f(0) + f(1)

In [29]: sum_karr(f(x), x, 2, 0)
Out[29]: -f(0) - f(1)

In [36]: sum_karr(f(x), x, 2, 0) == - sum_karr(f(x), x, 0, 2)
Out[36]: True
```
With the SymPy definition it looks like:
```python
In [34]: summation(f(x), (x, 0, 2-1))
Out[34]: f(0) + f(1)

In [35]: summation(f(x), (x, 2, 0-1))
Out[35]: -f(0) - f(1)
```

> the code-printer class should be fixed (so `lambdify` starts working).

Yes, the code printer and lambdify are wrong here and should be fixed to interpret the sum correctly.
